File: permute.py
class Solution:
    def generate(self, nums, curr, ret):
        if(len(nums) == 0):
            ret.append(curr)
            return
        else:
            for index in range(len(nums)): #Swap each index and permute
                nums[0], nums[index] = nums[index], nums[0]
                self.generate(nums[1:], curr + [nums[0]], ret)

    # Given a collection of distinct integers, return all possible permutations.
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = []
        self.generate(nums, [], ret)
        return ret

File: permute.js
/**
 Given a collection of distinct integers, return all possible permutations.
 * @param {number[]} nums
 * @return {number[][]}
 */

const permute = nums => {
    let ret = []

    const generate = (curr, remaining) => {
        curr = curr || []
        if(!remaining.length){
            ret.push(curr)
            return
        } else {
            //Move each index and permute
            for (let index = 0; index < remaining.length; index++){
                const currNum = remaining[index]
                generate(curr.concat(currNum), remaining.slice(0, index).concat(remaining.slice(index + 1)))
            }
        }
    }

    generate([], nums)
    return ret
}
